File: compiler/src/main/java/com/arch/jonnyhsia/compass/compiler/CompassProcessor.java
package com.arch.jonnyhsia.compass.compiler;

import com.arch.jonnyhsia.compass.facade.CompassPage;
import com.arch.jonnyhsia.compass.facade.ICompassTable;
import com.arch.jonnyhsia.compass.facade.annotation.Route;
import com.arch.jonnyhsia.compass.facade.TargetType;
import com.google.auto.service.AutoService;
import com.squareup.javapoet.AnnotationSpec;
import com.squareup.javapoet.ClassName;
import com.squareup.javapoet.JavaFile;
import com.squareup.javapoet.MethodSpec;
import com.squareup.javapoet.ParameterizedTypeName;
import com.squareup.javapoet.TypeSpec;

import java.io.IOException;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

import javax.annotation.processing.AbstractProcessor;
import javax.annotation.processing.Filer;
import javax.annotation.processing.Messager;
import javax.annotation.processing.ProcessingEnvironment;
import javax.annotation.processing.Processor;
import javax.annotation.processing.RoundEnvironment;
import javax.lang.model.SourceVersion;
import javax.lang.model.element.Element;
import javax.lang.model.element.ElementKind;
import javax.lang.model.element.Modifier;
import javax.lang.model.element.TypeElement;
import javax.lang.model.type.TypeMirror;
import javax.lang.model.util.Elements;
import javax.lang.model.util.Types;
import javax.tools.Diagnostic;

@AutoService(Processor.class)
public class CompassProcessor extends AbstractProcessor {

    private final static String TABLE_FULL_PATH = "compassTable";
    private final static String DEFAULT_SCHEME = "compassDefaultPageScheme";

    private Filer filer;
    private Messager messager;

    private String tablePackage;
    private String tableName;
    private String defaultScheme;

    private Elements elementUtils;
    private Types typeUtils;

    private TypeMirror typeActivity;
    private TypeMirror typeFragment;

    private int round = 0;

    @Override
    public synchronized void init(ProcessingEnvironment processingEnvironment) {
        super.init(processingEnvironment);
        filer = processingEnvironment.getFiler();
        messager = processingEnvironment.getMessager();
        elementUtils = processingEnvironment.getElementUtils();
        typeUtils = processingEnvironment.getTypeUtils();

        typeActivity = elementUtils.getTypeElement(Const.ACTIVITY).asType();
        typeFragment = elementUtils.getTypeElement(Const.FRAGMENT).asType();

        final String tableFullName = processingEnvironment.getOptions().get(TABLE_FULL_PATH);
        if (tableFullName != null && !tableFullName.isEmpty()) {
            final int indexOfLastPeriod = tableFullName.lastIndexOf(".");
            tablePackage = tableFullName.substring(0, indexOfLastPeriod);
            tableName = tableFullName.substring(indexOfLastPeriod + 1);
        } else {
            tablePackage = "com.arch.jonnyhsia.compass";
            tableName = "CompassTable";
        }

        final String scheme = processingEnvironment.getOptions().get(DEFAULT_SCHEME);
        if (scheme != null && !scheme.isEmpty()) {
            defaultScheme = scheme;
        } else {
            defaultScheme = "app";
        }
    }

    @Override
    public boolean process(Set<? extends TypeElement> annotations, RoundEnvironment roundEnvironment) {
        try {
            note("Hello from the other side...");
            // 跳过重复执行的 process
            if (annotations == null || annotations.isEmpty()) {
                note("No Annotations... Generate an empty table.");
                return true;
            }
            if (round > 0) {
                return false;
            }
            round++;

            List<RouteInfo> routeInfoList = new ArrayList<>();
            for (Element annotatedElement : roundEnvironment.getElementsAnnotatedWith(Route.class)) {
                if (annotatedElement.getKind() != ElementKind.CLASS) {
                    // 注解类型错误, 则跳过
                    error("该注解只能用于类", annotatedElement);
                    continue;
                }
                // 将注解信息保存在 RouteInfo 中, 并添加到数组
                routeInfoList.add(new RouteInfo((TypeElement) annotatedElement));
            }

            // 生成 Java Code
            generateJavaCode(routeInfoList);
        } catch (Exception e) {
            e.printStackTrace();
        }

        return true;
    }

    private void generateJavaCode(List<RouteInfo> routeInfoList) throws IOException {
        // HashMap<PageKey, CompassPage>
        ParameterizedTypeName pageMapType = ParameterizedTypeName.get(HashMap.class, String.class, CompassPage.class);

        // @Override
        // final Map<PageKey, CompassPage> loadPages()
        // HashMap<PageKey, CompassPage> map = new HashMap<>()
        MethodSpec.Builder getPagesMethod = MethodSpec.methodBuilder("getPages")
                .addAnnotation(AnnotationSpec.builder(Override.class).build())
                .addModifiers(Modifier.FINAL, Modifier.PUBLIC)
                .returns(pageMapType)
                .addStatement("$T map = new $T<>()", pageMapType, HashMap.class);

        // map.put(PageKey, CompassPage)
        for (RouteInfo routeInfo : routeInfoList) {
            TargetType targetType = getTargetTypeByRouteInfo(routeInfo);
            getPagesMethod.addComment(routeInfo.getRouteString());
            getPagesMethod.addCode("map.put($S, new $T($S, $T.class, $L, $T.$L, $S",
                    routeInfo.getName(),
                    CompassPage.class,
                    routeInfo.getName(), routeInfo.getTarget(), routeInfo.getRequestCode(),
                    targetType.getClass(), targetType.name(),
                    pageScheme(routeInfo.getScheme()));

            if (routeInfo.getInterceptors().size() == 0) {
                getPagesMethod.addCode(", new $T[0]", Class.class);
            } else {
                getPagesMethod.addCode(", new $T[] {", Class.class);
                for (int i = 0; i < routeInfo.getInterceptors().size(); i++) {
                    ClassName interceptorClz = routeInfo.getInterceptors().get(i);
                    getPagesMethod.addCode("$T.class", interceptorClz);
                    if (i != routeInfo.getInterceptors().size() - 1) {
                        getPagesMethod.addCode(", ");
                    }
                }
                getPagesMethod.addCode("}");
            }

            getPagesMethod.addStatement("));");
        }
        // return map
        getPagesMethod.addStatement("return map");

        // public final class XXX { }
        TypeSpec tableClass = TypeSpec.classBuilder(tableName)
                .addModifiers(Modifier.PUBLIC, Modifier.FINAL)
                .addSuperinterface(ICompassTable.class)
                .addMethod(getPagesMethod.build())
                .build();

        // com.arch.jonnyhsia.compass.XXX
        JavaFile javaFile = JavaFile.builder(tablePackage, tableClass).build();
        javaFile.writeTo(filer);
    }

    private TargetType getTargetTypeByRouteInfo(final RouteInfo routeInfo) {
        if (typeUtils.isSubtype(routeInfo.getType(), typeActivity)) {
            return TargetType.ACTIVITY;
        } else if (typeUtils.isSubtype(routeInfo.getType(), typeFragment)){
            return TargetType.FRAGMENT;
        }
        return TargetType.UNKNOWN;
    }

    private String pageScheme(String scheme) {
        return scheme.isEmpty() ? defaultScheme : scheme;
    }

    private void note(String format, Object... args) {
        messager.printMessage(Diagnostic.Kind.NOTE, String.format(format, args));
    }

    private void error(CharSequence charSequence, Element element) {
        messager.printMessage(Diagnostic.Kind.ERROR, charSequence, element);
    }

    @Override
    public Set<String> getSupportedAnnotationTypes() {
        return new HashSet<String>() {{
            add(Route.class.getCanonicalName());
        }};
    }

    @Override
    public SourceVersion getSupportedSourceVersion() {
        return SourceVersion.latest();
    }
}

File: compiler/src/main/java/com/arch/jonnyhsia/compass/compiler/CompassProcessor.kt
package com.arch.jonnyhsia.compass.compiler

import com.arch.jonnyhsia.compass.api.*
import com.google.devtools.ksp.KspExperimental
import com.google.devtools.ksp.getClassDeclarationByName
import com.google.devtools.ksp.processing.Dependencies
import com.google.devtools.ksp.processing.Resolver
import com.google.devtools.ksp.processing.SymbolProcessor
import com.google.devtools.ksp.processing.SymbolProcessorEnvironment
import com.google.devtools.ksp.symbol.KSAnnotated
import com.google.devtools.ksp.symbol.KSClassDeclaration
import com.google.devtools.ksp.symbol.KSType
import com.squareup.kotlinpoet.*
import com.squareup.kotlinpoet.ParameterizedTypeName.Companion.parameterizedBy
import com.squareup.kotlinpoet.ksp.KotlinPoetKspPreview
import kotlin.concurrent.thread

@KotlinPoetKspPreview
@KspExperimental
class CompassProcessor(
    environment: SymbolProcessorEnvironment
) : SymbolProcessor {

    private companion object {
        val ROUTE_NAME = Route::class.qualifiedName!!
        const val TABLE_FULL_PATH = "compassTable"
    }

    private val codeGenerator = environment.codeGenerator
    private val logger = environment.logger

    private val tablePackage: String
    private val tableName: String

    private val routeSymbols = ArrayList<KspRouteSymbol>()

    private lateinit var dependencies: Dependencies

    init {
        val tableFullName = environment.options[TABLE_FULL_PATH]
        if (!tableFullName.isNullOrEmpty()) {
            val indexOfLastPeriod = tableFullName.lastIndexOf(".")
            tablePackage = tableFullName.substring(0, indexOfLastPeriod)
            tableName = tableFullName.substring(indexOfLastPeriod + 1)
        } else {
            tablePackage = "com.arch.jonnyhsia.compass"
            tableName = "CompassTable"
        }
    }

    private lateinit var activityType: KSType
    private lateinit var fragmentType: KSType

    override fun process(resolver: Resolver): List<KSAnnotated> {
        activityType = resolver.getClassDeclarationByName("android.app.Activity")!!.asType()
        fragmentType =
            resolver.getClassDeclarationByName("androidx.fragment.app.Fragment")!!.asType()

        logger.warn("Start processing...")
        logger.warn("getSymbolsWithAnnotation: $ROUTE_NAME")
        val symbols = resolver.getSymbolsWithAnnotation(ROUTE_NAME)

        dependencies = Dependencies(false, *resolver.getAllFiles().toList().toTypedArray())

        symbols.forEach { type ->
            if (type is KSClassDeclaration) {
                routeSymbols.add(KspRouteSymbol(type))
            } else {
                logger.warn("$type is not class declaration")
            }
        }
        logger.warn("Count: ${routeSymbols.size}")
        return emptyList()
    }

    override fun finish() {
        generateCode(routeSymbols)
        super.finish()
    }

    private fun generateCode(annotations: List<KspRouteSymbol>) {
        thread {
            val filename = tableName
            codeGenerator.createNewFile(
                dependencies = dependencies,
                packageName = tablePackage,
                fileName = filename,
                extensionName = "kt"
            ).writer().use { writer ->
                val tableType = HashMap::class.asClassName().parameterizedBy(
                    PageKey::class.asTypeName(),
                    CompassPage::class.asTypeName()
                )
                logger.warn("hello: $annotations")
                FileSpec.builder(tablePackage, filename)
                    .addType(
                        TypeSpec.objectBuilder(filename)
                            .addSuperinterface(ICompassTable::class)
                            .addFunction(
                                FunSpec.builder("getPages")
                                    .addModifiers(KModifier.OVERRIDE)
                                    .returns(tableType)
                                    .collectAllRoutesAndReturn(annotations, tableType)
                                    .build()
                            )
                            .build()
                    )
                    .build()
                    .writeTo(writer)
            }
        }
    }

    private fun FunSpec.Builder.collectAllRoutesAndReturn(
        symbols: List<KspRouteSymbol>,
        tableType: ParameterizedTypeName
    ): FunSpec.Builder {
        logger.warn("hello222")
        addComment("Read routes(${symbols.size}) from annotations and generate router map")
        addStatement("val map = %T()", tableType)

        for (symbol in symbols) {
            val isActivity = activityType.isAssignableFrom(symbol.targetKsType)
            val isFragment = fragmentType.isAssignableFrom(symbol.targetKsType)
//            logger.warn("${symbol.symbol.asType().toClassName().packageName} isAssignableFrom: $result")
            logger.warn("Activity isAssignableFrom symbol: $isActivity")
            logger.warn("Fragment isAssignableFrom symbol: $isFragment")

            val targetType: TargetType = when {
                isActivity -> TargetType.ACTIVITY
                isFragment -> TargetType.FRAGMENT
                else -> TargetType.UNKNOWN
            }
            addStatement(
                "map[%T(%S)] = %T(%S, %T::class.java, %L, %T.%L, %S)",
                PageKey::class.java,
                symbol.route.name,
                CompassPage::class.java,
                symbol.route.name,
                symbol.target,
                symbol.route.requestCode,
                targetType::class.java, targetType.name,
                symbol.route.scheme
            )
        }
        addCode("return map")
        return this
    }
//    inner class RouteVisitor : KSVisitorVoid() {
//        override fun visitClassDeclaration(classDeclaration: KSClassDeclaration, data: Unit) {
//            super.visitClassDeclaration(classDeclaration, data)
//        }
//    }
}
